docs/design/copy-tracking.md

```diff
@@ -10,17 +10,613 @@ results of rebasing across remote copies.
+for the log and annotation use cases.
+
+Should we support copy tracking for symlinks? Their history is not very for
+annotation purposes, but knowing the history may at least be useful for
+detecting directory renames (if all files and symlinks in a directory were
+renamed).
+
+We won't support tracking copied directories.
+
+The data structure might look like this:
+```rust
+// Current `TreeValue::File` variant:
+File { id: FileId, executable: bool },
+// New `TreeValue::File` variant:
+File { id: FileId, executable: bool, copy_id: CopyId },
+
+// A CopyId is a hash of this struct:
+struct CopyHistory {
+    path: RepoPath,
+    parents: Vec<CopyId>
+}
+```
+
+### Diffing
+
+When diffing two trees, we first diff the trees without considering copy info.
+For any copy IDs that changed in that diff, we walk all of their copy graphs to
+figure out how they're related and which source file to associate with which
+destination file.
+
+The details of the algorithm is left for the implementation. The following
+sections provide some examples to hopefully show that it's feasible.
+
+#### Example: Divergent copy and rename
+
+Let's look at an example of how this model would look in this scenario:
+
+```
+M rename foo->baz, create bar
+|
+| L copy foo->bar, create baz
+|/
+K add foo
+```
+
+Assuming the new files are different in each commit, we get these trees:
+```
+Commit K:
+name: foo, id: K, copy_id: 1:foo
+
+Commit L:
+name: bar, id: L, copy_id: 2:bar->1:foo
+name: baz, id: L, copy_id: 3:baz
+name: foo, id: K, copy_id: 1:foo
+
+Commit M:
+name: bar, id: M, copy_id: 4:bar
+name: baz, id: M, copy_id: 5:baz->1:foo
+```
+
+The `2:bar->1:foo` syntax above means that copy ID 2 has file `bar`, which was
+copied from copy ID `1`, where it was called `foo`.
+
+Diffing from `L` to `M` finds 2,3,4,5 as changed copy IDs. By walking their
+graphs, we find that 1,2, and 5 are related, while 3 and 4 are not.
+
+Because the `bar` and `baz` files have unrelated copy graphs, we break up their
+diffs into two separate diffs for each file.
+
+Among the remaining copy IDs, the shortest path in the copy graph is between
+`foo` on the source side and `baz` on the destination side, so we start with.
+Since `foo` doesn't exist on the source side and `baz` doesn't exist on the
+destination side (with a related copy ID), we consider it a rename.
+
+The remaining file is `bar` on the source side. Its closest relative on the
+destination side is `baz`. Since we already used `baz` as a rename target for
+`foo`, we won't consider `bar` renamed to it. So we consider `bar` as copied
+into `baz`.
+
+So we get these diffs:
+
+* `baz` is deleted (deleting content `L`)
+* `bar` is created (with content `M`)
+* `foo` is renamed to `baz` (showing diff from `K` to `M`)
+* `bar` is merged into `baz` (showing diff from `L` to `M`)
+
+#### Example: Divergent copy and rename (best rename target)
+
+```
+N copy baz->qux
+|
+M rename bar->baz
+|
+| L rename foo->bar
+|/
+K add foo
```

Comment:
nit: maybe something missing in `K` or between `K` and `M`? Since `L` renames to `bar`, `bar` doesn't exist in `L`'s parent `K`.

Reply:
Perhaps M could rename foo->baz?